### PR TITLE
Add luminibble unpacker

### PR DESCRIPTION
**Pull request by capalmer85**
This module works and is safe, but it does not get into the Events tree.  How should we go about that?

### CODE REVIEW

**Comment by mommsen** [comment]
+1

**Comment by davidlange6** [comment]
@capalmer85 - thinking ahead - could we give this producer a more generic name - such as LumiRawToDigi (both in C++ and python)?

To save the product, you need to add it to Configuration/EventContent (you should make a new cff following the examples in the existing cffs)

**Comment by capalmer85** [comment]
@davidlange6 OK.  I'll change the labels and sort out the event content today.

File: EventFilter/LumiRawToDigi/python/Lumi_EventContent_cff.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import FWCore.ParameterSet.Config as cms
+#
+# Author: Chris Palmer
+# Date:   02.06.15
+# Notes:  
+#
+
+LumiEventContent = cms.PSet(
+    outputCommands = cms.untracked.vstring(
+        'keep *'
```
Comment (capalmer85) [inline EventFilter/LumiRawToDigi/python/Lumi_EventContent_cff.py]: I'm not sure what I should put here.

**Comment by capalmer85** [comment]
When I run only LumiRawToDigi (EDProducer) and an OuputModule, the branch comes through as "int_myLabel_nibble_myprocessname".  However, when I run the full machinery, I can see that my module is running (I print the nibble), but it doesn't make it to the output.  I'm not sure why it isn't going to the event content.  Can any of you see from the code?  

(for reference)
/afs/cern.ch/user/c/capalmer/public/NibbleProducer/simple_cfg.py
/afs/cern.ch/user/c/capalmer/public/NibbleProducer/full_cfg.py

File: EventFilter/LumiRawToDigi/python/Lumi_EventContent_cff.py (hunk 1 above)
Comment (slava77) [inline EventFilter/LumiRawToDigi/python/Lumi_EventContent_cff.py]: it should be smth like ``` keep *_lumiDigis_*_* ``` if we want to keep everything.

**Comment by slava77** [comment]
@davidlange6 what was the rationale to turn this into a more generic LumiRawToDigi with this implementation already ?

We typically unpack by FED/raw data type rather by the purpose of the data.
This one started as a partial  TCDS unpacker.
If a more generic name for the producer was needed, probably a TCDSRawToDigi is more appropriate.

All other lumi-related info in FEDRawData was coming in scalers FED and is currently unpacked in the ScalersRawToDigi.
As I understand some of it may change with the BRIL DAQ inputs available, which may or may not end up in the scalers.
If they show up as a different FED data block, then that's something to be put in LumiRawToDigi.

What do you think?

**Comment by davidlange6** [comment]
@slava77 - i'd be happy with TCDSraw->digi too. Up to you.

(but actually putting this code into the scalers package is also a good idea - i'd rather not start a new package for this)

**Comment by slava77** [comment]
@davidlange6 
My suggestion was to stay with TCDS raw to digi  in the same EventFilter/Utilities where the current event header making based on TCDS FED data is already happening. 
This is currently purely in DAQ area, but I can request for the reconstruction to be added here as well.

For the name of the producer, TcdsRawToDigi is probably a more appropriate name given our style conventions, but TCDSRawToDigi is OK as well.

**Comment by capalmer85** [comment]
Ok.  So I will change Lumi to Tcds in the name of the modules and put them into EventFilter/Utilities as follows:

EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc  
-->  EventFilter/Utilities/plugins/TcdsRawToDigi.cc

EventFilter/LumiRawToDigi/python/LumiRawToDigi_cfi.py
-->  EventFilter/Utilities/python/TcdsRawToDigi_cfi.py

EventFilter/LumiRawToDigi/python//Lumi_EventContent_cff.py
-->  EventFilter/Utilities/python//Tcds_EventContent_cff.py


Can someone help me figure out why my nibble isn't getting into the event content?

File: EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        nibble=-1;
+    }
+    std::cout<<"nibble is "<<nibble<<std::endl;
+
+    std::unique_ptr<int> pOut(new int(nibble));
+    iEvent.put( std::move(pOut), "nibble" );
+}
+
+// ------------ method called once each job just before starting event loop  ------------
+void
```
Comment (slava77) [inline EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc]: remove these (beginJob and endJob)

File: EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+
+
+using namespace std;
+
+//
+// class declaration
+//
+
+class LumiRawToDigi : public edm::EDProducer {
```
Comment (slava77) [inline EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc]: change to ```stream::EDProducer<>```

**Comment by slava77** [comment]
On 6/4/15 9:17 AM, capalmer85 wrote:
> Ok. So I will change Lumi to Tcds in the name of the modules and put
> them into EventFilter/Utilities as follows:
> 
> EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc
> 
> --> EventFilter/Utilities/plugins/TcdsRawToDigi.cc
> 
> EventFilter/LumiRawToDigi/python/LumiRawToDigi_cfi.py
> --> EventFilter/Utilities/python/TcdsRawToDigi_cfi.py

since you implemented the fillDescriptions,
we can use the system-generated default cfi
tcdsRawToDigi_cfi.py

So, the TcdsRawToDigi_cfi is superfluous and can be removed.

> 
> EventFilter/LumiRawToDigi/python//Lumi_EventContent_cff.py
> --> EventFilter/Utilities/python//Tcds_EventContent_cff.py
> 
> Can someone help me figure out why my nibble isn't getting into the
> event content?

visually, it should work.
Please update the PR to be checked there.

		--slava

File: EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc (hunk 3 above)
Comment (capalmer85) [inline EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc]: class LumiRawToDigi : public edm::EDProducer { 
to
class LumiRawToDigi : public stream::EDProducer<> {
causes a compiler error.

 error: 'stream' has not been declared
 class LumiRawToDigi : public stream::EDProducer<> {

Do I need to add an include?

Comment (slava77) [inline EventFilter/LumiRawToDigi/plugins/LumiRawToDigi.cc]: yes ```/EDProducer.h > /stream/EDProducer.h ```

**Comment by capalmer85** [comment]
Updates complete.  nibble is still not in event content.

File: EventFilter/Utilities/plugins/TcsdRawToDigi.cc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+   public:
+      explicit TcdsRawToDigi(const edm::ParameterSet&);
+      ~TcdsRawToDigi();
+
+      static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+   private:
+      virtual void produce(edm::Event&, const edm::EventSetup&) override;
+      
+      int nibble;
```
Comment (slava77) [inline EventFilter/Utilities/plugins/TcsdRawToDigi.cc]: please use a local variable inside the ::produce method implementation.
There is no need to keep it here.

File: EventFilter/Utilities/plugins/TcsdRawToDigi.cc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        if(tcdsData.size()>0){
+            evf::evtn::TCDSRecord tcdsRecord(tcdsData.data());
+            nibble = (int)tcdsRecord.getHeader().getData().header.nibble;
+        } else {
+            nibble=-2;
+        }
+    } else {
+        nibble=-1;
+    }
+    std::cout<<"nibble is "<<nibble<<std::endl;
```
Comment (slava77) [inline EventFilter/Utilities/plugins/TcsdRawToDigi.cc]: change to LogTrace or remove.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by capalmer85** [comment]
I was using an event content in my cfg that I hadn't added the TcdsEventContent to.  I've fixed that and a typo in a filename (Tcsd --> Tcds).  I think this is finally done now.

**Pull request merged**